File: 백준/Gold/9251. LCS/LCS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		String n = br.readLine().trim();
		String m = br.readLine().trim();

		int[][] memo = new int[n.length() + 1][m.length() + 1];

		for (int i = 1; i <= n.length(); i++) {
			for (int j = 1; j <= m.length(); j++) {
				if (n.charAt(i - 1) == m.charAt(j - 1)) {
					memo[i][j] = memo[i - 1][j - 1] + 1;
				} else {
					memo[i][j] = Math.max(memo[i - 1][j], memo[i][j - 1]);
				}
			}
		}

		System.out.println(memo[n.length()][m.length()]);
	}
}

File: 백준/Gold/9251. LCS/LCS.py
import sys

n = sys.stdin.readline().rstrip()
m = sys.stdin.readline().rstrip()

ln = len(n)
lm = len(m)

memo = [[0 for _ in range(lm + 1)] for _ in range(ln + 1)]

for i in range(1, ln + 1):
    for j in range(1, lm + 1):
        if n[i - 1] == m[j - 1]:
            memo[i][j] = memo[i - 1][j - 1] + 1
        else:
            memo[i][j] = max(memo[i - 1][j], memo[i][j - 1])

print(memo[ln][lm])
